nixos/tests/nextcloud/with-objectstore.nix

```diff
@@ -0,0 +1,97 @@
+  adminpass = "hunter2";
+  adminuser = "root";
+
+  accessKey = "nextcloud";
+  secretKey = "test12345";
+
+  rootCredentialsFile = pkgs.writeText "minio-credentials-full" ''
+    MINIO_ROOT_USER=${accessKey}
+    MINIO_ROOT_PASSWORD=${secretKey}
+  '';
+
+in {
+  name = "nextcloud-with-objectstore";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ onny ];
+  };
+
+  nodes = {
+    # The only thing the client needs to do is download a file.
+    client = { ... }: {};
+
+    nextcloud = { config, pkgs, ... }: {
+      networking.firewall.allowedTCPPorts = [ 80 ];
+
+      services.nextcloud = {
+        enable = true;
+        hostName = "nextcloud";
+        https = true;
+        package = pkgs.${"nextcloud" + (toString nextcloudVersion)};
+        database.createLocally = true;
+        # Don't try this at home!
+        config.adminpassFile = "${pkgs.writeText "adminpass" adminpass}";
+        config.objectstore.s3 = {
+          enable = true;
+          bucket = "nextcloud";
+          autocreate = true;
+          key = "nextcloud";
+          secretFile = "${pkgs.writeText "secretKey" secretKey}";
+          hostname = "localhost";
+          port = 9000;
+          usePathStyle = true;
+          region = "us-east-1";
+        };
+      };
+
+      services.minio = {
+        enable = true;
+        listenAddress = "127.0.0.1:9000";
+        consoleAddress = "127.0.0.1:9001";
+        inherit rootCredentialsFile;
+      };
+
+    };
+  };
+
+  testScript = let
+    withRcloneEnv = pkgs.writeScript "with-rclone-env" ''
+      #!${pkgs.runtimeShell}
+      export RCLONE_CONFIG_NEXTCLOUD_TYPE=webdav
+      export RCLONE_CONFIG_NEXTCLOUD_URL="http://nextcloud/remote.php/dav/files/${adminuser}"
+      export RCLONE_CONFIG_NEXTCLOUD_VENDOR="nextcloud"
+      export RCLONE_CONFIG_NEXTCLOUD_USER="${adminuser}"
+      export RCLONE_CONFIG_NEXTCLOUD_PASS="$(${pkgs.rclone}/bin/rclone obscure ${adminpass})"
+    '';
+    copySharedFile = pkgs.writeScript "copy-shared-file" ''
+      #!${pkgs.runtimeShell}
+      echo 'hi' | ${pkgs.rclone}/bin/rclone rcat nextcloud:test-shared-file
+    '';
+
+    diffSharedFile = pkgs.writeScript "diff-shared-file" ''
+      #!${pkgs.runtimeShell}
+      diff <(echo 'hi') <(${pkgs.rclone}/bin/rclone cat nextcloud:test-shared-file)
+    '';
+  in ''
+    start_all()
+
+    nextcloud.wait_for_unit("multi-user.target")
+    nextcloud.wait_for_unit("minio.service")
+    nextcloud.wait_for_open_port(9000)
+    nextcloud.succeed(
+      "${pkgs.minio-client}/bin/mc config host add nextcloud http://localhost:9000 ${accessKey} ${secretKey} --api s3v4"
+    )
+    nextcloud.succeed("${pkgs.minio-client}/bin/mc mb minio/nextcloud")
+
+    nextcloud.succeed("curl -sSf http://nextcloud/login")
+    nextcloud.succeed(
+        "${withRcloneEnv} ${copySharedFile}"
+    )
+    client.wait_for_unit("multi-user.target")
+    client.succeed(
+        "${withRcloneEnv} ${diffSharedFile}"
+    )
```

Comment:
I think it would also be nice to ensure that the data is actually in the S3 bucket (should be doable with `awscli`). Otherwise there's a risk that we miss some regressions where Nextcloud fails to connect to S3 at all.

Reply:
i tried to use minio-client "mc" which works fine in a virtual machine but somehow not in the test environment :( somehow this is really complicated ....